Add XPath library for parsing xpath config paths

tsyesika opened the pull request:
This library adds an library which parses a small subset of xpath queries against a yang schema. The library currently supports querying arrays (specified as `position()=1` for the first element) based on the position, querying lists based on keys (specified as `[addr=1.2.3.4]`) and specifying leaves and containers based on their name (more information in documentation at the top of the file).

The path is transformed into a lua path (as suggested in #549). The only difference is that instead of just being a table of keys or just being an integer it's a table with a "name" field and a key or keys field. This change is because our query is XPath and not the adhoc query language which specified paths sligthly differently.

teknico committed:
Add explanation of more Hydra jobset parameters (#563)

chrgraf committed:
added howto craft packets via scapy tool

kbara committed:
Merge pull request #575 from chrgraf/lwaftr

added howto craft packets via scapy tool

Commit:
Add XPath parser for xpath -> lua path

This converts an XPath query which would be made against a yang
model and converts it to a lua path. This path can the be used
to easily resolve a node in a data tree.

Commit:
Remove unnecessary imports in selftest

teknico committed:
Add CSV output to the lwaftr loadtest command (#578)

wingo commented on src/lib/yang/xpath.lua:
```diff
@@ -0,0 +1,146 @@
+-- The structure of the path is dependent on the type the node is, the
+-- conversions are as follows:
+--
+-- Scala fields:
```
"Scalar"

wingo commented on src/lib/yang/xpath.lua:
```diff
@@ -0,0 +1,146 @@
+-- named keys such as [addr=1.2.3.4] and also basic positional querying
+-- for arrays e.g [position()=1] for the first element.
+--
+-- The structure of the path is dependent on the type the node is, the
```
This is a comma splice, please use full stop or semicolon

wingo replied:
lol my comment was also a comma splice

wingo commented on src/lib/yang/xpath.lua:
```diff
@@ -0,0 +1,146 @@
+-- -- Use of this source code is governed by the Apache 2.0 license; see COPYING.
+
+-- This module can be used to parse a path based on a yang schema (or it's
```
its

wingo commented on src/lib/yang/xpath.lua:
```diff
@@ -0,0 +1,146 @@
+-- Loads a module and converts the rest of the path.
+function load_from_path(path)
+   -- First extract and load the module name then load it.
+   module_name, path = next_element(fragment)
```
needs local on these vars, and fragment seems undefined

wingo commented on src/lib/yang/xpath.lua:
```diff
@@ -0,0 +1,146 @@
+local function extract_parts(fragment)
+   local rtn = {query={}}
+   rtn.name = string.match(fragment, "([^%[]+)")
+   for k,v in string.gmatch(fragment, "%[([^=]+)=([^%]]+)%]") do
```
u r a wizzard

wingo commented on src/lib/yang/xpath.lua:
```diff
@@ -0,0 +1,146 @@
+end
+function handlers.table(fragment, tree)
+   if #tree.keys == 1 then
+      -- Check if the key is a string based key
```
There is a "string_key" property on tables which are string-keyed

wingo replied:
this will fail for tables that have 1 key but are not string-keyed.  the test for tree.string_key is sufficient

wingo commented on src/lib/yang/xpath.lua:
```diff
@@ -0,0 +1,146 @@
+end
+
+-- Converts an XPath path to a lua array consisting of path componants.
+-- A path compent can then be resolved on a yang data tree:
```
component

wingo reviewed (commented):
Looking very good, some nits.  Can we add a facility to apply a path to a data tree?  I guess it would take the parsed path and return a function that could apply to data.  I suppose there would be a general version that would return an array of results and a version built on that that would return just one result or nil, or something like that.

wingo commented:
Ping :)  I would love to deliver this tomorrow!

teknico committed:
Add Hydra mode to the lwaftr loadtest command (#583)

Add Hydra mode to the lwaftr loadtest command

wingo committed:
Merge pull request #560 from Igalia/yang

lwAFTR uses new YANG-based configuration

wingo committed:
Version 3.0.0

teknico committed:
Add missing IMIX pcap files (#584)

Commit:
Fix nits and add resolve to lib.yang.xpath

This fixes some spelling issues and some other small nits brought up in
review. This also adds a resolve function which takes a schema, data tree
and path and resolves the path for the data tree and corrisponding schema.

xray7224 commented:
Pushed fixes for everything you brought up. I also added a resolve command and extra tests for it.

PTAL

Commit:
Fix invalid library references in lib.yang.xpath

wingo commented on src/lib/yang/xpath.lua:
```diff
@@ -0,0 +1,281 @@
+-- A path component can then be resolved on a yang data tree:
+
+local function convert_path(path, grammar)
+   -- Extract head, check type and dispatch to handler.
```
Can we instead break these passes up:
1. lex the path into components by splitting on /
2. transform each component by mapping that split path to parsed path components
Here they are a bit twingled and I find it difficult to read.

wingo commented on src/lib/yang/xpath.lua:
```diff
@@ -0,0 +1,281 @@
+   local node
+   if grammar.type == "table" then
+      if grammar.keys[head] == nil then
+         node = assert(grammar.values[parts.name], err)
```
Do we want to support looking up entries in a table by value?  I would guess not

tsyesika replied:
This is to support looking inside of lists, to use an example from the tests "/routes/route[addr=1.2.3.4]/port" it would be the "port". We don't support list selection via values, only by keys.

wingo commented on src/lib/yang/xpath.lua:
```diff
@@ -0,0 +1,281 @@
+   end
+end
+
+function resolve(schema, data, path)
```
Can we turn this into 

```
function resolver(schema, path)
   return function(data) ... end
end
```

i.e. compute the resolver without having the data?  Then we can apply that same resolver to many data items.

Commit:
Add on-a-stick mixed IPv4 IPv6 benchdata pcap

Originally by tekniko, checksums fixed and filenames changed as per discussion

kbara committed:
Merge pull request #587 from Igalia/stick_imix_pcap

Add on-a-stick mixed IPv4 IPv6 benchdata pcap

Commit:
Refactor lib.yang.xpath.convert_path and fix nits

This completely refactors the convert_path function responsible for taking
a yang xpath path and converting it to a lua path that can be used to resolve
a node in a yang data tree. The refactor in general makes the code more readable
by splitting some of the functionality out of the convert_path function into smaller,
better defined functions.

This also changes the return value from the resolve function to be a function which
takes a data tree and resolves for that data tree. This allows the resolver to be
re-used on multiple data trees with the same schema and path.

Commit:
Fix final nits in lib.yang.xpath

xray7224 commented:
I think I've fixed all the nits, please take another look :)

wingo commented on src/lib/yang/xpath.lua:
```diff
@@ -0,0 +1,283 @@
+         return grammar.keys[fragment.name]
+      end
+   else
+      return assert(grammar[fragment.name], errmsg)
```
I think this is probably wrong -- you want e.g. grammar.members for a struct.  WDYT?  I would assert in this function that type is either table or struct.  Or at least that it's not scalar.

wingo replied:
AFAIU this comment still applies

wingo commented on src/lib/yang/xpath.lua:
```diff
@@ -0,0 +1,283 @@
+end
+
+-- Returns a resolver for a paticular schema and *lua* path.
+function resolve(schema, path)
```
Consider renaming this to `resolver`, as it just computes the resolver and doesn't actually run the thing

wingo commented on src/lib/yang/xpath.lua:
```diff
@@ -0,0 +1,283 @@
+
+-- Returns a resolver for a paticular schema and *lua* path.
+function resolve(schema, path)
+   local schema = lib.deepcopy(schema)
```
No need to do this unless you are mutating the schema and I don't thing we should be mutating the schema; wdyt?

wingo commented on src/lib/yang/xpath.lua:
```diff
@@ -0,0 +1,283 @@
+   local ret = {}
+   local node = grammar
+   for element in split_path(path) do
+      node = extract_grammar_node(node, element)
```
Very nice function.  WDYT about extracting the grammar node and annotating the lua path with that node?  That changes the Lua representation of paths of course (each element becomes a table with some well-known keys) but given that we always have to compile paths to functions (getters, setters, augmenters, removers) then having this information handy can make these compilation procedures easier to write.

wingo commented on src/lib/yang/xpath.lua:
```diff
@@ -0,0 +1,283 @@
+      prod = prod.values[peek]
+      return handle(scm, prod, data, path)
+   end
+   function handlers.container(scm, prod, data, path)
```
I think here we should be operating on the grammar only and not on the schema.  WDYT?

tsyesika replied:
So for the handlers in the resolving code I think we need to operate on the schema to be able to find the primitive type in the list handler for example. The other handles otherwise don't really use it and just advance it on.

I can't think of a way to re-write the list handler without the schema but i'm open to suggestions if I've missed something?

Commit:
Fix some more nits in lib.yang.xpath brought up by review

xray7224 commented:
I've fixed those nits (even though the review doesn't show them all as being addressed, they have been). Let me know what you think @wingo

wingo commented on src/lib/yang/xpath.lua:
```diff
@@ -0,0 +1,280 @@
+      if #path == 0 then return data end
+      return assert(handlers[scm.kind], scm.kind)(scm, prod, data, path)
+   end
+   function handlers.list(scm, prod, data, path)
```
I don't think we should be interpreting the schema here -- instead we should be building a resolver based on the grammar.  The grammar describes the concrete data format whereas the schema is at a higher level (e.g. presence), and the path should be in terms of data/grammar, not schema.  This is reflected in that the YANG XPath expressions in RFC 6020 apply (AFAIU) to the XML data, which follows the XML mapping rules.  (For us the grammar is like what happens to the schema after you apply XML mapping rules).  So we should have cases for tables, scalars, and so on, not containers, leaves, etc.  WDYT?

wingo commented on src/lib/yang/xpath.lua:
```diff
@@ -0,0 +1,280 @@
+      return handle(scm, prod, data, path)
+   end
+   return function (data)
+      local data = lib.deepcopy(data)
```
Why copy the data? :)

Commit:
Tighten up definitions in lib.yang.xpath

This refactors a bunch of code to tighten up the definitions of functions
that are in lib.yang.xpath. These do more checking, are easier to read and
more versatile. I've removed arguments which didn't *have* to be there.

wingo commented on src/lib/yang/xpath.lua:
```diff
@@ -0,0 +1,260 @@
+function convert_path(grammar, path)
+   local ret = {}
+   local node = grammar
+   for element in split_path(path) do
```
Could `for element in path:split('/')` work here?  There is a definition of `string:split` in `core.lib`.

wingo commented on src/lib/yang/xpath.lua:
```diff
@@ -0,0 +1,260 @@
+      if #path ~= 0 then error("Paths can't go beyond leaves.") end
+      return data[normalize_id(head.name)]
+   end
+   return function (data)
```
Can this be expressed in such a way that the handler is computed ahead of time?  I.e. do all of the lookups in `handlers` before returning the function.

wingo committed:
Rename xpath.lua to path.lua

wingo committed:
Localize handlers for path.lua:convert_path

Commit:
Fix a bug and use string:split() in lib.yang.xpath

wingo committed:
lib.yang.path test tweaks

wingo committed:
Merge 'lwaftr' into xpath

wingo committed:
Precompute path resolvers before running

Before, we would look at the grammar when resolving a path in data, to
determine what to do at the next path item.  Now we resolve what to do
ahead of time, and just return a closure that does the lookup.

Resolve also returns the grammar for the data that will be looked up, as
a second value.

wingo commented:
In order for path lookup not to call ffi.typeof all the time, I wanted to use data.typeof which was only added later.  Also we need to make this PR against lwaftr, not yang.  For this reason I am closing then opening this PR.
